pkg/controllers/cassandra/nodepool/nodepool.go

```diff
@@ -0,0 +1,131 @@
+package nodepool
+
+import (
+	"fmt"
+
+	v1alpha1 "github.com/jetstack/navigator/pkg/apis/navigator/v1alpha1"
+	"github.com/jetstack/navigator/pkg/controllers/cassandra/util"
+	apps "k8s.io/api/apps/v1beta1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	appslisters "k8s.io/client-go/listers/apps/v1beta1"
+	"k8s.io/client-go/tools/record"
+)
+
+type Interface interface {
+	Sync(*v1alpha1.CassandraCluster) error
+}
+
+type defaultCassandraClusterNodepoolControl struct {
+	kubeClient        kubernetes.Interface
+	statefulsetLister appslisters.StatefulSetLister
+	recorder          record.EventRecorder
+}
+
+var _ Interface = &defaultCassandraClusterNodepoolControl{}
+
+func NewControl(
+	kubeClient kubernetes.Interface,
+	statefulsetLister appslisters.StatefulSetLister,
+	recorder record.EventRecorder,
+) Interface {
+	return &defaultCassandraClusterNodepoolControl{
+		kubeClient:        kubeClient,
+		statefulsetLister: statefulsetLister,
+		recorder:          recorder,
+	}
+}
+
+func ownerCheck(
+	set *apps.StatefulSet,
+	cluster *v1alpha1.CassandraCluster,
+) error {
+	if !metav1.IsControlledBy(set, cluster) {
+		ownerRef := metav1.GetControllerOf(set)
+		return fmt.Errorf(
+			"Foreign owned StatefulSet: "+
+				"A StatefulSet with name '%s/%s' already exists, "+
+				"but it is controlled by '%v', not '%s/%s'.",
+			set.Namespace, set.Name, ownerRef,
+			cluster.Namespace, cluster.Name,
+		)
+	}
+	return nil
+}
+
+func (e *defaultCassandraClusterNodepoolControl) removeUnusedStatefulSets(
+	cluster *v1alpha1.CassandraCluster,
+) error {
+	expectedStatefulSetNames := map[string]bool{}
+	for _, pool := range cluster.Spec.NodePools {
+		name := util.NodePoolResourceName(cluster, &pool)
+		expectedStatefulSetNames[name] = true
+	}
+	client := e.kubeClient.AppsV1beta1().StatefulSets(cluster.Namespace)
+	selector, err := util.SelectorForCluster(cluster)
+	if err != nil {
+		return err
+	}
+	existingSets, err := e.statefulsetLister.
+		StatefulSets(cluster.Namespace).
+		List(selector)
+	if err != nil {
+		return err
+	}
+	for _, set := range existingSets {
+		err := ownerCheck(set, cluster)
+		if err != nil {
+			return err
+		}
```

Comment:
Does it make sense to return an error here, if a cluster is found with matching labels but the ownerRef is wrong? This will cause the rest of the node pool synchronisation to fail (and also cause the sync to be retried)

If it does make sense to return an error here and have it retried later, we should log an Event to the CassandraCluster resource to make it clear that a StatefulSet matching labels but not owned by blah blah blah exists and cannot be modified (and recommending the course of action to a user).

That said, I'd have thought we're best to just ignore that StatefulSet and perhaps log a warning that we came across something that matches labels, but isn't owned.

Reply:
Ah - I've realised that the error returned from `ownerCheck` is already quite descriptive. My only remaining point then is just verifying that this should cause this sync to be aborted and retried later.

Reply:
I thought it safest to err here after vaguely  thinking of the case where user creates > deletes > recreates a cluster with the same name.
Until we implement  garbage collect of the the old resources we'll run into this problem, I think.

Reply:
Garbage collection wouldn't prevent this though, as it would only garbage collect things that are 'owned'. This code basically says "if a StatefulSet that has the label X=Y on it exists, and it's not owned by this CassCluster, this is an error state and we cannot proceed".

This makes sense to me *iff* that StatefulSet *also* has the same name as another statefulset that we *do* want to own (because we wouldn't want to overwrite an existing deployment). Right now though, we just bail out even if it would be possible to proceed from what I can see.

In the case where a user creates->deletes->creates, the check for ownerRef [here](https://github.com/jetstack/navigator/pull/59/files/f05f42e03efdaaae8dec1ff34c7b30dea96c597a#diff-9c2ba35894d4eb5cf3a1e42b156eac33R106) should pick it up and bail out (when the actual create happens).

Reply:
Will lgtm just to keep this moving, but I do think we should revisit this clause 😄